File: 프로그래머스/3/12920. 선입 선출 스케줄링/선입 선출 스케줄링.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

int solution(int n, vector<int> cores) {
    if(n <= cores.size()) return n;
    n -= cores.size();
    
    int left = 0;
    int right = n * *max_element(cores.begin(), cores.end());
    int time = right;
    while(left <= right){
        int mid = (left + right) / 2;
        int sum = accumulate(cores.begin(), cores.end(), 0, [&mid](int acc, int now){ return acc + mid / now; });
        
        if(n <= sum){
            right = mid - 1;
            time = mid;
        }
        else{
            left = mid + 1;
        }
    }
    int answer = 0;
    int left_work = n - accumulate(cores.begin(), cores.end(), 0, [&time](int acc, int now){ return acc + (time - 1) / now; });
    for(int i = 0; i < cores.size(); i++){
        if(time % cores[i] == 0){
            left_work--;
            if(left_work == 0){
                answer = i + 1;
                break;
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/3/12920. 선입 선출 스케줄링/선입 선출 스케줄링.py
def solution(n, cores):
    if n <= len(cores): return n
    
    n -= len(cores)
    left = 0
    right = max(cores) * n
    time = 0
    while left <= right:
        mid = (left + right) // 2
        s = sum(mid // core for core in cores)
        
        if n <= s:
            time = mid
            right = mid - 1
        else:
            left = mid + 1
    
    remain_work = n - sum((time - 1) // core for core in cores)
    answer = 0
    for i, core in enumerate(cores):
        if time % core == 0:
            remain_work -= 1
            
            if remain_work == 0:
                answer = i + 1
                break
            
    return answer